### PR TITLE
Fix poll transmit and min mtu

### PR DESCRIPTION
**Pull request by stormshield-gt**
This PR adds a test for post-quantum key exchange. For now the test is failing to demonstrate the presence of a bug; I will then propose a fix.

Note that I added a commit that exposed the key exchange group, like this was done for cipher in https://github.com/quinn-rs/quinn/pull/2001

### CODE REVIEW

File: quinn-proto/src/crypto/rustls.rs [hunk 1]
```diff
@@ -263,6 +268,8 @@ pub struct HandshakeData {
     pub server_name: Option<String>,
     /// The ciphersuite negotiated with the peer
     pub negotiated_cipher_suite: CipherSuite,
+    /// The key exchange group negotiated with the peer
+    pub negociated_key_exchange_group: NamedGroup,
```
Comment (djc) [inline quinn-proto/src/crypto/rustls.rs:272]: `negociated` -> `negotiated`, please.

File: quinn/tests/post_quantum.rs [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+use rustls::{
+    pki_types::{CertificateDer, PrivatePkcs8KeyDer},
+    NamedGroup,
+};
+
+use std::{
+    error::Error,
+    net::{Ipv4Addr, SocketAddr},
+    sync::Arc,
+};
```
Comment (djc) [inline quinn/tests/post_quantum.rs:16]: Please sort imports in `StdExternalCrate` blocks where both `quinn` and `proto` count as part of the crate.

File: quinn/tests/post_quantum.rs [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+    let (endpoint, server_cert) = make_server_endpoint(server_addr)?;
+    // accept a single connection
+    tokio::spawn(async move {
+        let incoming_conn = endpoint.accept().await.unwrap();
+        let conn = incoming_conn.await.unwrap();
+        println!(
+            "[server] connection accepted: addr={}",
+            conn.remote_address()
+        );
```
Comment (djc) [inline quinn/tests/post_quantum.rs:31]: Use tracing events instead of `println!()` (and set up the usual subscriber).

File: quinn/tests/post_quantum.rs [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    });
+
+    let endpoint = make_client_endpoint((Ipv4Addr::UNSPECIFIED, 0).into(), &[&server_cert])?;
+    // connect to server
+    let connection = endpoint
+        .connect(server_addr, "localhost")
+        .unwrap()
+        .await
+        .unwrap();
+    println!("[client] connected: addr={}", connection.remote_address());
```
Comment (djc) [inline quinn/tests/post_quantum.rs:50]: Tracing.

File: quinn-proto/src/crypto/rustls.rs (hunk 1 above)
Comment (stormshield-gt) [reply quinn-proto/src/crypto/rustls.rs]: oups, thanks for spoting this

File: quinn/tests/post_quantum.rs (hunk 3 above)
Comment (stormshield-gt) [reply quinn/tests/post_quantum.rs]: I've done this; I've also added the `with_test_writer` to the `subscriber` because I noticed otherwise the log weren't printed.

File: quinn/tests/post_quantum.rs (hunk 4 above)
Comment (stormshield-gt) [reply quinn/tests/post_quantum.rs]: I've changed it. In general, do you prefer I close the thread or you do it?

File: quinn/tests/post_quantum.rs (hunk 2 above)
Comment (stormshield-gt) [reply quinn/tests/post_quantum.rs]: thanks that's cleaner, I've done it

File: quinn/tests/post_quantum.rs (hunk 4 above)
Comment (djc) [reply quinn/tests/post_quantum.rs]: If you're reasonably sure you've addressed the feedback, feel free to resolve the thread.

**Comment by Ralith** [comment]
https://github.com/quinn-rs/quinn/pull/2027 gets this test passing.

**Comment by stormshield-gt** [comment]
I've rebased and split into 2 tests, one for the `substract with overflow` error and one for the `debug_assert`

**Comment by stormshield-gt** [comment]
The freebsd job seems to failed because `cmake` is not installed on the runner

File: quinn-proto/src/crypto/rustls.rs [hunk 5]
```diff
@@ -11,7 +11,7 @@ use rustls::{
     client::danger::ServerCertVerifier,
     pki_types::{CertificateDer, PrivateKeyDer, ServerName},
     quic::{Connection, HeaderProtectionKey, KeyChange, PacketKey, Secrets, Suite, Version},
-    CipherSuite,
+    CipherSuite, NamedGroup,
```
Comment (djc) [inline quinn-proto/src/crypto/rustls.rs:14]: Tiny nit: can you also fix the "negociated" in the commit message?

File: quinn/tests/post_quantum.rs [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+use rustls::{
+    pki_types::{CertificateDer, PrivatePkcs8KeyDer},
+    NamedGroup,
+};
+use std::{
+    error::Error,
+    net::{Ipv4Addr, SocketAddr},
+    sync::Arc,
+};
+use tracing::info;
```
Comment (djc) [inline quinn/tests/post_quantum.rs:15]: Nit: please use separate blocks for `std`, external, and `quinn` imports (in that order).

File: quinn/tests/post_quantum.rs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    // Waiting for a stream will complete with an error when the server closes the connection
+    let _ = connection.accept_uni().await;
+
+    // Make sure the server has a chance to clean up
+    endpoint.wait_idle().await;
+}
+
+fn make_client_endpoint(
+    bind_addr: SocketAddr,
+    server_certs: &[&[u8]],
```
Comment (djc) [inline quinn/tests/post_quantum.rs:75]: Let's do a single `server_cert: CertificateDer<'_>` instead, to match what we actually need to pass in here?

**Comment by djc** [comment]
> The freebsd job seems to failed because `cmake` is not installed on the runner

Why does this PR change the build requirements? Can you change the job to install `cmake`?

**Comment by stormshield-gt** [comment]
That's because `rustls-post-quantum` is not optional and unconditionally brings `aws-lc-rs`. Sadly, it's not possible to have an optional dev dependency currently in `cargo`. 

I used a "private" feature hack to work around this, like they do, for instance, in the `reqwest` crate. Please let me know if you have a better solution.

File: quinn-proto/src/crypto/rustls.rs [hunk 8]
```diff
@@ -69,6 +69,11 @@ impl crypto::Session for TlsSession {
                 .negotiated_cipher_suite()
                 .expect("cipher is negotiated")
                 .suite(),
+            negotiated_key_exchange_group: self
+                .inner
+                .negotiated_key_exchange_group()
+                .expect("key exchange group is negociated")
```
Comment (BiagioFesta) [inline quinn-proto/src/crypto/rustls.rs:75]: typo: `negociated`

Comment (stormshield-gt) [reply quinn-proto/src/crypto/rustls.rs:75]: Thanks for catching this

**Comment by stormshield-gt** [comment]
I've discovered a new corner case with `min_mtu` of 1433,
with the current feature setup (that avoids compiling `aws-lc-rs` on freebsd and android), the command to test locally is

```rust
RUST_LOG=trace cargo t --test post_quantum -F __rustls-post-quantum-test -- --show-output
```

**Comment by Ralith** [comment]
Fixed the new case in #2046.

**Comment by stormshield-gt** [comment]
Thanks again for the fixes, I can confirm it works also on my computer

File: quinn-proto/src/crypto/rustls.rs [hunk 9]
```diff
@@ -256,6 +261,8 @@ pub struct HandshakeData {
     ///
     /// Always `None` for outgoing connections
     pub server_name: Option<String>,
+    /// The key exchange group negotiated with the peer
+    pub negotiated_key_exchange_group: NamedGroup,
```
Comment (djc) [inline quinn-proto/src/crypto/rustls.rs:265]: Sorry, we don't want to land this on main right now due to semver compatibility issues.

Comment (stormshield-gt) [reply quinn-proto/src/crypto/rustls.rs:265]: Which option do you prefer: 
- Remove the commit that adds this. In that case I will only test that the handshake succeeds, not the specific key exchange group.
- Keep this PR open until the next breaking release

Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:265]: First option sounds okay to me, I'll shortly reopen a PR to add back the API and then we can tack on a commit to add this back to the test as well?

Comment (stormshield-gt) [reply quinn-proto/src/crypto/rustls.rs:265]: @stormshield-damiend suggested, as we use a `__rustls-post-quantum-test` for testing, maybe I can declare this field under this feature flag?

Comment (djc) [reply quinn-proto/src/crypto/rustls.rs:265]: Ohh, that makes sense to me, yes!

Comment (stormshield-gt) [reply quinn-proto/src/crypto/rustls.rs:267]: I did this in my last push :tada:

File: quinn/tests/post_quantum.rs [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+};
+use tracing::info;
+
+use quinn::{
+    crypto::rustls::{HandshakeData, QuicClientConfig, QuicServerConfig},
+    Endpoint,
+};
+
+#[tokio::test]
+async fn post_quantum_key_worse_case_header() {
```
Comment (djc) [inline quinn/tests/post_quantum.rs:21]: `worse_case` -> `worst_case`?

File: quinn/tests/post_quantum.rs [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    check_post_quantum_key_exchange(1433, 8082).await;
+}
+
+async fn check_post_quantum_key_exchange(min_mtu: u16, server_port: u16) {
+    let _ = tracing_subscriber::FmtSubscriber::builder()
+        .with_env_filter(tracing_subscriber::EnvFilter::from_default_env())
+        .with_test_writer()
+        .try_init();
+
+    let _ = rustls_post_quantum::provider().install_default();
```
Comment (djc) [inline quinn/tests/post_quantum.rs:36]: Would prefer we pass this into the `rustls::ClientConfig` and `rustls::ServerConfig` builders directly.

File: quinn/tests/post_quantum.rs [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+    let server_addr = (Ipv4Addr::LOCALHOST, server_port).into();
```
Comment (djc) [inline quinn/tests/post_quantum.rs:37]: Nit: spell this as `SocketAddr::from((Ipv4Addr::LOCALHOST, server_port));`?

**Review by djc — APPROVED** [review]

**Comment by djc** [comment]
Thanks!

**Comment by djc** [comment]
@Ralith want to review this? Should be straightforward. @stormshield-gt up for a rebase?

**Comment by stormshield-gt** [comment]
From what I understand, when activating `aws-lc-rs` feature of `rustls` 0.24, PQ key exchange would be the default thanks to https://github.com/rustls/rustls/pull/2552

For now I think I will use the `prefer-post-quantum` feature of `rustls` 0.23

**Comment by djc** [comment]
> From what I understand, when activating `aws-lc-rs` feature of `rustls` 0.24, PQ key exchange would be the default thanks to [rustls/rustls#2552](https://github.com/rustls/rustls/pull/2552)
> 
> For now I think I will use the `prefer-post-quantum` feature of `rustls` 0.23

`prefer-post-quantum` was enabled by default in [0.23.27](https://github.com/rustls/rustls/releases/tag/v%2F0.23.27) so I think you don't need to use it here (since our `Cargo.lock` has a newer version already).

**Comment by stormshield-gt** [comment]
Thanks for spotting this, I don't enable the feature explicitly anymore.

File: quinn/tests/post_quantum.rs [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+use tracing::info;
+
+use quinn::{
+    Endpoint,
+    crypto::rustls::{HandshakeData, QuicClientConfig, QuicServerConfig},
+};
+
+#[tokio::test]
+async fn post_quantum_key_worst_case_header() {
+    check_post_quantum_key_exchange(1274, 8081).await;
```
Comment (Ralith) [inline quinn/tests/post_quantum.rs]: nit: Using hard-coded port numbers is prone to spurious failure, since that port may be in used by an unrelated service on the system. A more robust approach is to specify port 0 (wildcard) when binding a `UdpSocket`, then query the actual `local_addr` it ends up with, allowing the OS to allocate an arbitrary free port.

**Review by Ralith — APPROVED** [review]

File: Cargo.toml [hunk 14]
```diff
@@ -42,6 +42,7 @@ rustls = { version = "0.23.5", default-features = false, features = ["std"] }
 rustls-pemfile = "2"
 rustls-platform-verifier = "0.6"
 rustls-pki-types = "1.7"
+rustls-post-quantum = "0.1"
```
Comment (Ralith) [inline Cargo.toml]: I'm a bit confused by the mechanism here. This dependency doesn't seem to actually be referenced by anything.

Comment (djc) [reply Cargo.toml]: Yeah, it should no longer be used AIUI.

Comment (stormshield-gt) [reply Cargo.toml]: Yep, I forgot to remote it when I used the `prefered-post-quantum` feature

File: quinn/tests/post_quantum.rs (hunk 13 above)
Comment (stormshield-gt) [reply quinn/tests/post_quantum.rs]: Thanks, that's a lot better
